### PR TITLE
Add test engine to manage snapshot verification testing

### PR DESCRIPTION
**Pull request by redgoat650**
Test engine manages the test and metadata repositories, snapshot
checker, metadata storage persistence, and file writer. It is
the high level helper that will be invoked in the snapshot
verification testing suite.

- modify data directory file structure
- issue snapshot/restore/delete to the data directory
- accumulate metadata over the course of the test suite
- flush accumulated metadata to the metadata repository
- load historical metadata from the repository on initialization
- perform automatic data integrity verification on snap restore

### CODE REVIEW

**Comment by redgoat650** [comment]
Note: this PR will the **last** of a series that will eventually converge to https://github.com/kopia/kopia/compare/master...kastenhq:test-engine . See that comparison for the full context of how and where this change will be used.

Complementary PRs:
- #40 - Comparer interface [fswalker] for data integrity verification
- #41 - Snapshotter interface [kopia] for issuing snapshots to repository
- #43 - Storer and DataPersister [kopia] interfaces for managing and persisting snapshot integrity metadata
- #44 - Checker that manages the interaction of the above interfaces for taking/restoring snapshots
- #45 - (This PR) Engine that manages metadata persistence, file writing [fio], repository connection/creation, and is the high-level tool that will be invoked to manage and run robustness testing.

**Comment by redgoat650** [comment]
Still TODO:
- Fix this issue for CI (needs `S3_BUCKET_NAME` env or skip test, or else will always fail).
- File writer interface
- File writer diversity (adding directory writing, better helpers etc)
- Integrating engine tests into CI
- Writing robustness test suite

**Comment by redgoat650** [comment]
As noted above, this PR is expected to fail in the current CI environment. Will fix it on Monday.

File: tests/robustness/test_engine/engine.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+// is successful, the engine is populated with the metadata associated with the
+// snapshot in that repo. A new repo will be created if one does not already
+// exist.
+func (e *Engine) InitS3(ctx context.Context, testRepoPath, metaRepoPath string) error {
+	bucketName := os.Getenv(S3BucketNameEnvKey)
+	if bucketName == "" {
+		return ErrS3BucketNameEnvUnset
+	}
+
+	err := e.MetaStore.ConnectOrCreateS3(bucketName, metaRepoPath)
```
Comment (julio-lopez) [inline tests/robustness/test_engine/engine.go]: Does it have to be S3? Can this be made blob store agnostic?

IIRC given the appropriate configuration, kopia would create the bucket if it does not exist.

**Comment by julio-lopez** [comment]
Overall, LGTM.
It'd be good to have a high-level walkthrough, it can be offline.

File: tests/robustness/test_engine/engine.go (hunk 1 above)
Comment (redgoat650) [reply tests/robustness/test_engine/engine.go]: Are you asking specifically about the metadata store repo? Or just connection to any repository?

The assumption and limitation of these two initialization helper functions (`eng.InitS3()` and `eng.InitFilesystem()`) is that the caller will connect/create a metadata repo that's in the same blob store provider as the test repo. For example if we want to connect to a filesystem test repo, we assume we'll find the metadata repo also in filesystem. Same for S3, where additionally the init function expects the metadata to live in the same bucket.

That said, these are all exported fields and methods, so if an Engine user wants to customize it where they have the metadata repo and test repo living in different buckets, or in different blob store services altogether, they can call `ConnectorCreate<blobstore>()` for each with the appropriate parameters.

These `ConnectOrCreate` functions are part of the `RepoManager` interface, which any repository (kopia, restic, whatever) must implement to be managed by the test engine (or at the very least return an error). Currently only the S3 and Filesystem funcs are added to the interface, but more can be added, to tell the engine to look for repos in google or wherever else.

Comment (redgoat650) [reply tests/robustness/test_engine/engine.go]: oh also

>IIRC given the appropriate configuration, kopia would create the bucket if it does not exist.

I don't think it does, unless I'm missing some required setup:

```
kopia: error: can't connect to storage: bucket "nick-kasten-io-test-2" does not exist, try --help
```

**Pull request merged**